### DISCUSSION TITLE
Service Changed Characteristic

### DISCUSSION THREAD

**Discussion by scochet** [Moddable-OpenSource/moddable#837] [Q&A] [answered]
Hi, I am having a really good time programming ESP32 BLE with Moddable, and I need a way to tell the client that the device name has changed, since LightBlue on iOS (indeed, iOS) uses the first GAP device name found and cached.

I think that the service changed characteristic would be enough, but it is not clear (to me) how to do it in Moddable.
May you help me, hopefully through an example?

Thanks,
Daniele

**Comment by andycarle** [comment]
Hi Daniele,

The Moddable SDK does not have any special support for the service changed characteristic. You should, however, be able to deploy that characteristic exactly the same as any other GATT characteristic by using a [GATT Service JSON file](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/network/ble/ble.md#gattservices).

Simply create a GATT Service JSON File that defines the Generic Attribute Service (UUID `0x1801`) with a Service Changed characteristic (UUID `0x2A05`). (I looked up those UUIDs in [this document](https://btprodspecificationrefs.blob.core.windows.net/assigned-values/16-bit%20UUID%20Numbers%20Document.pdf)). Then notify on that characteristic when your service has changed. See the [Heart Rate Server](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/network/ble/heart-rate-server) example for an example of how to notify on characteristic changes.

That said, I'm not actually sure what iOS's behavior is when you indicate a Service Change. It may or may not pick up the device name change — that's an implementation-specific detail.

Please give this a shot and let me know how it goes. If you have trouble setting up the right GATT service, I can try putting together a little example later today.

&nbsp;- Andy

**Reply by scochet** [reply]
Hi Andy,

thank you for your kind reply. I tried to write some code according to your suggestions, resulting in this (which is not working anyway) in a gattrs.json:

```json
{
	"service": {
		"uuid": "1801",
		"characteristics": {
			"service_changed": {
				"uuid": "2A05",
                "type": "Array",
                "maxBytes": 20,
				"permissions": "read",
				"properties": "indicate",
				"value": ["0000", "FFFF"]
			}
		}
	}
}
```

and somwhere in my .js code: 

```
this.notifyValue({"uuid":uuid`2A05`,"name":"service_changed","type":"Array","notify":1}, [0x0000, 0xFFFF]);
```

I can't manage to tell the client (i.e. iOS) that the device name changed, it remains the same...

**Comment by scochet** [comment, marked as answer]
Hi, 

I came up with this. I suspect that something is redundant and not-so-correct or elegant, but it is working so far, and useful.

gap.json:
```json
{
	"service": {
		"uuid": "1800",
		"characteristics": {
			"device_name": {
				"uuid": "2A00",
				"maxBytes": 16,
				"type": "String",
				"permissions": "read",
				"properties": "read",
				"value": "My (new) device!"
			},
			"appearance": {
				"uuid": "2A01",
				"maxBytes": 2,
				"type": "Uint16",
				"permissions": "read",
				"properties": "read",
				"value": 1362
			}
		}
	}
}
```

gattrs.json:
```json
{
	"service": {
		"uuid": "1801",
		"characteristics": {
			"service_changed": {
				"uuid": "2A05",
                "type": "Array",
                "maxBytes": 20,
				"permissions": "read",
				"properties": "indicate",
				"value": ["0000", "FFFF"]
			}
		}
	}
}
```

This way, as soon as iOS connects to the device, it finds the service changed characteristic and modifies values in cache accordingly.

**Comment by phoddie** [comment]
@scochet - this looks like something other developers would benefit from. It would be a great example for the Moddable SDK. Would you be willing to share a working example? A PR would be great. If it is easier for you to share in this discussion, that's fine too.

> ...I suspect that something is redundant and not-so-correct or elegant, but it is working so far, and useful.

With a working example, it would be straightforward for @andycarle to take a look to see if he has any suggestions to offer.

**Reply by andycarle** [reply]
Indeed, @scochet, I'd be happy to take a look at your code in full and see if I have any advice. 

Having a working example of this process in [the BLE examples folder](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/network/ble) would be very helpful to other developers. If you have a minimal example that could live there, a PR would be the way to proceed. Otherwise, if you can share your code here I can spin up an example based off of it.

Thanks!
